### DISCUSSION TITLE
How to use Hybrid properties which return select objects

### DISCUSSION THREAD

**Discussion by gioxc88** [sqlalchemy/sqlalchemy#5945] [Usage Questions]
I have a simple class which represents a Portfolio with financial assets and I want to define two hybrid properties,
The `amount_local` which is just `price * quantity`  and the `amount` which is `amount_local / fx`. 

```
from datetime import datetime
from sqlalchemy import Column, Integer, Float, ForeignKey, case, func
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property

Base = declarative_base()

class PortfolioComposition(Base):
    date = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    id_portfolio = Column(ForeignKey('portfolio.id'), primary_key=True, index=True)  # input
    id_asset = Column(ForeignKey('asset.id'), primary_key=True, index=True, nullable=False)
    quantity = Column(Integer)
    price = Column(Float)
    accrued = Column(Float)
    fx = Column(Float)

    @hybrid_property  # calculated
    def amount_local(self):
        return self.quantity * (self.price + (self.accrued or 0)) / getattr(self.asset, 'nominal', 1)

    @amount_local.expression
    def amount_local(cls):
        q = session.query(case(
            [
                (Asset.category == 'bond', cls.quantity * (cls.price + cls.accrued) / Bond.nominal)
            ],
            else_=cls.quantity * cls.price
        )).select_from(cls).outerjoin(Bond)
        return q.statement

    @hybrid_property  # calculated
    def amount(self):
        return self.amount_local / self.fx

    @amount.expression  # calculated
    def amount(cls):
        return cls.amount / cls.fx
```

Now the for the `amount_local` I defined both `hybrid_property` and `expression` and it works fine if I just query the column on its own like 

`session.query(PortfolioComposition.amount_local).all()`

however the following doesn't work

`session.query(PortfolioComposition.amount).all()`

if fails with 

```
sqlalchemy.exc.StatementError: (builtins.TypeError) float() argument must be a string or a number, not 'Select'
[SQL: SELECT ? / portfolio_composition_live.fx AS amount 
FROM portfolio_composition_live]
[parameters: [immutabledict({})]]
```

and also the following doesn't work

`session.query(func.sum(PortfolioComposition.amount_local))`

if fails with 

`sqlalchemy.exc.InvalidRequestError: Select objects don't have a type.`

Does anyone have any suggestion on how to approach this usecase?
I've run out of ideas.

Many thanks
Gio

**Comment by zzzeek** [comment]
it looks like you should be converting your select() to a scalar subquery:

    # ...
    return stmt.as_scalar()

Also I would recommend using a select() construct directly rather than query(), as that way you don't need a "Session" which you are throwing away in any case.    The session.query() pattern becomes legacy in SQLAlchemy 1.4 overall.

1.3 version:

```
   from sqlalchemy.orm import join as orm_join
   @amount_local.expression
    def amount_local(cls):
        return select([
            case(
            [
                (Asset.category == 'bond', cls.quantity * (cls.price + cls.accrued) / Bond.nominal)
            ],
            else_=cls.quantity * cls.price
        )
        ]).select_from(orm_join(cls, Bond)).as_scalar()


```
1.4 version:
```

   @amount_local.expression
    def amount_local(cls):
        return select(
            case(
                (Asset.category == 'bond', cls.quantity * (cls.price + cls.accrued) / Bond.nominal),
            else_=cls.quantity * cls.price
        )
        ).select_from(cls).join(Bond).scalar_subquery()

```

it is hoped that 1.4 is the more obvious style as we've tried to get rid of a lot of old patterns that get in the way.

**Reply by gioxc88** [reply]
Thank you very much for the answer I tried to use `as_scalar` before but the problem is that instead of returning all the records it returns only the first one. I am not sure why, do you any any clue?


for example if I use `as_scalar` at the end I get

```
session.query(PortfolioComposition.id_asset, PortfolioComposition.quantity, PortfolioComposition.amount_local).all()

Out[6]: 
[(5, -888274.92, -888274.92),
 (338, 100000, -888274.92),
 (374, 2200, -888274.92),
 (376, 5000, -888274.92),
 (377, 17000, -888274.92),
 (378, 2200, -888274.92),
 (379, 6900, -888274.92),
...
```
which is incorrect because it returns only the first record associated to the `hybrid_property` `amount_local`

If I remove `as_scalar`  instead all the rows are duplicated as if it were a carthesian product of the last two columns
@zzzeek  Any idea why `as_scalar` behaves like this?

Thank you
Gio

**Reply by zzzeek** [reply]
[scalar subqueries](https://mariadb.com/kb/en/subqueries-scalar-subqueries/) by definition return only one column and one row.   If you are looking to emit SQL that is going to JOIN to another table and result in more rows being returned, then you need to spell that out in your query, outside of the scope of a "hybrid" attribute.   hybrid attributes only resolve to SQL column expressions, not SELECTs that aren't scalar.

for me to help here I would need to see the SQL statement that you hope to generate.

**Reply by gioxc88** [reply]
I am working on a minimal reproducible example.
But I really want a very simple thing:

```
AssetCategory
id    name
1    equity
2     bond

Asset
id id_category
1        1
2        1
3        2
```
`amount` is `hybrid_property` => IF AssetCategory.name = 'bond' THEN price * quantity /100 ELSE price * quantity
`weight` is a `hybrid_property` which depend on `amount` hence I want to use `amount` inside other `hibrid_property.expression`
`weight` is the the percentage amount for each id_portfolio

If I call the sum the amounts per id_portfolio `total_1` and `total_2`
In the example below
`total_1 = 20 + 45+ 16/100` 
`total_2 = 60 + 21/100`
```
PortfolioComposition
id_portfolio id_asset price quantity amonut       weight
       1          1    10      2       20       20/total_1
       1          2     9      5       45       45/total_!
       1          3     8      2     16/100        ...

       2          1    10      6       60       60/total_2
       2          3     7      3     21/100        ...

AssetCategory(Base)
    id = Column(Integer, primary_key=True) 
    name = Column(String(50))

Asset(Base)
    id = Column(Integer, primary_key=True) 
    id_category = Column(ForeignKey('asset_category.id'), nullable=False)
    
    category = relationship('AssetCategory')

PortfolioComposition(Base)
    id_portfolio = Column(ForeignKey('portfolio.id'), primary_key=True, index=True) 
    id_asset = Column(ForeignKey('asset.id'), primary_key=True, index=True, nullable=False)
    quantity = Column(Integer)
    price = Column(Float)

    asset = relationship('Asset')

    @hybrid_property  
    def amount(self):
        return self.price * self.quantity / 100 if self.asset.category.name == 'bond' else self.price * self.quantity 

    @amount.expression 
    def amount(cls):
        ...
```

So my questions are: 
1. how can a `hybrid_property` return only a single scalar? I have many rows in my `PortfolioComposition` table so I expect the `hybrid_property` to return a scalar per each row?
2. In your opinon is `hybrid_property` the right approach for achieving my goal?
3. In case the answer to 2 is yes do would you have some code to share on how to to this that can serve as a guide?

Finally I managed to make it work using `where` clause instead of `join`  (as in the first post) but I honestly don't understand the logic behind it (and why it works that way) so I would really appreciate if you could make light on it.
Just to show how I made it work I'll post the code

```
@amount.expression
    def amount(cls):
        return select([func.round(case(
            [
                (Asset.id_asset_category == 2, cls.quantity * cls.price  / 100)
            ],
            else_=cls.quantity * cls.price
        ), 5)]).where(Asset.id == cls.id_asset).label('amount_local')


@weight.expression
def weight(cls):
    sub_query = select([cls.id_portfolio.label('ptf'), func.sum(cls.amount).label('group_sum')]).group_by(cls.id_portfolio)
    return select([cls.amount / sub_query.c.group_sum]).where(and_(
            cls.id_portfolio == sub_query.c.ptf,
            Asset.id == cls.id_asset
        )
    ).label('weight')
```



Many thanks
Gio

**Reply by zzzeek** [reply]
> I am working on a minimal reproducible example.

I'm known for demanding these from people, but in this case really just the SQL string you are looking for would be sufficient for me to understand what you want.    you should always have an idea of the SQL structure you are looking to represent when having issues like these; trying to connect Python things together without a picture of SQL syntax is very difficult.


> But I really want a very simple thing:
> 
> ```
> AssetCategory
> id    name
> 1    equity
> 2     bond
> 
> Asset
> id id_category
> 1        1
> 2        1
> 3        2
> ```
> 
> `amount` is `hybrid_property` => IF AssetCategory.name = 'bond' THEN price * quantity /100 ELSE price * quantity
> `weight` is a `hybrid_property` which depend on `amount` hence I want to use `amount` inside other `hibrid_property.expression`
> `weight` is the the percentage amount for each id_portfolio

so that's a CASE statement, which you have.   It seems like perhaps you want those other tables, assetcategory, etc. in the FROM clause, and likely using a JOIN of some kind.   That's fine, but the hybrid_property can't do that automatically, you would need to include those tables in the enclosing query, and the hybrid returns just the case() expression.

> 
> So my questions are:
> 
>     1. how can a `hybrid_property` return only a single scalar? I have many rows in my `PortfolioComposition` table so I expect the `hybrid_property` to return a scalar per each row?

to be accurate, hybrid_property at the `@expression` level returns a SQL expression, that is then rendered inside of a SELECT statement passed to the database.   it does not return a "value".    this is how you need to be thinking to use this attribute effectively.  


> 
>     2. In your opinon is `hybrid_property` the right approach for achieving my goal?

I would need to see the SQL string but it sounds like an issue that is extremely common, which is people want to SELECT from a combination of tables and they are seeking to find some way for these tables to automatically apply themselves to the FROM clause, which is not a feature that a hybrid_property provides by itself.     

> 
>     3. In case the answer to 2 is yes do would you have some code to share on how to to this that can serve as a guide?

if you can supply the SQL string you are looking to emit, then yes.
